src/runtime_asan.js

```diff
@@ -8,111 +8,22 @@
 #error "should only be inclded in USE_ASAN mode"
 #endif
 
-// C versions of asan_js_{load|store}_* will be used from compiled code, which have
+// C versions of asan_js_{load|store} will be used from compiled code, which have
 // ASan instrumentation on them. However, until the wasm module is ready, we
 // must access things directly.
 
-/** @suppress{duplicate} */
-function _asan_js_load_1(ptr) {
-  if (runtimeInitialized) return __asan_c_load_1(ptr);
-  return HEAP8[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_1u(ptr) {
-  if (runtimeInitialized) return __asan_c_load_1u(ptr);
-  return HEAPU8[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_2(ptr) {
-  if (runtimeInitialized) return __asan_c_load_2(ptr);
-  return HEAP16[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_2u(ptr) {
-  if (runtimeInitialized) return __asan_c_load_2u(ptr);
-  return HEAPU16[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_4(ptr) {
-  if (runtimeInitialized) return __asan_c_load_4(ptr);
-  return HEAP32[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_4u(ptr) {
-  if (runtimeInitialized) return __asan_c_load_4u(ptr) >>> 0;
-  return HEAPU32[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_8(ptr) {
-  if (runtimeInitialized) return __asan_c_load_8(ptr);
-  return HEAP64[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_8u(ptr) {
-  if (runtimeInitialized) return BigInt.asUintN(64, __asan_c_load_8u(ptr));
-  return HEAPU64[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_f(ptr) {
-  if (runtimeInitialized) return __asan_c_load_f(ptr);
-  return HEAPF32[ptr];
-}
-/** @suppress{duplicate} */
-function _asan_js_load_d(ptr) {
-  if (runtimeInitialized) return __asan_c_load_d(ptr);
-  return HEAPF64[ptr];
+function _asan_js_load(arr, index) {
+  if (runtimeInitialized) {
+    const elemSize = arr.BYTES_PER_ELEMENT;
+    ___asan_loadN(index * elemSize, elemSize);
+  }
```

Comment:
How does this handle signed/unsigned differences?

Reply:
See above, asan_loadN is a `void` function that only needs to check if pointer is valid and aborts if not. The beauty of this approach is that it doesn't need to care about signed/unsigned differences - instead, we just retrieve value with correct sign by retrieving it from the provided array (`HEAP32` / `HEAPU32` / whatever).

Reply:
Oh nice, now I see. Yes, very simple.